specs/agents/metrics-otel.md

```diff
@@ -0,0 +1,181 @@
+
+However, agents MUST support explicit bucket histogram aggregations with custom boundaries provided by users via views.
+
+Before export, the OpenTelemetry histogram bucket data needs to be converted to IntakeV2 histogram data. Because the OpenTelemetry specification allows negative values in histograms, the data needs to be converted to T-Digests: For T-Digests, the IntakeV2 metricset `samples.values` properties holds the **midpoints** of the histogram buckets instead of the upper/lower bounds.
+
+Therefore, the following algorithm needs to be used to convert [OpenTelemetry explicit bucket histogram data](https://opentelemetry.io/docs/reference/specification/metrics/data-model/#histogram) to IntakeV2 histograms:
+
+```
+IntakeV2Histo convertBucketBoundaries(OtelHisto input) {
+    List<Double> intakeCounts = new ArrayList<>();
+    List<Double> intakeValues = new ArrayList<>();
+
+    List<Double> otelCounts = input.getCounts();
+    List<Double> otelBoundaries = input.getBoundaries();
+    
+    int bucketCount = otelCounts.size();
+    // otelBoundaries has a size of bucketCount-1
+    // the first bucket has the boundaries ( -inf, otelBoundaries[0] ]
+    // the second bucket has the boundaries ( otelBoundaries[0], otelBoundaries[1] ]
+    // ..
+    // the last bucket has the boundaries (otelBoundaries[bucketCount-2], inf)
+    
+    for (int i = 0; i < bucketCount; i++) {
+        if (otelCounts.get(i) != 0) { //ignore empty buckets
+            intakeCounts.add(otelCounts.get(i));
+            if (i == 0) { //first bucket
+                double bounds = otelBoundaries.get(0);
+                if(bounds > 0) {
+                    bounds /= 2;
+                }
+                intakeValues.add(bounds);
+            } else if (i == bucketCount - 1) { //last bucket
+                intakeValues.add(otelBoundaries.get(bucketCount - 2));
+            } else { //in between
+                double lower = otelBoundaries.get(i - 1);
+                double upper = otelBoundaries.get(i);
+                intakeValues.add(lower + (upper - lower) / 2);
+            }
+        }
+    }
+    
+    return new IntakeV2Histo(intakeCounts, intakeValues);
+}
+```
+
+The same algorithm is used by the APM server to convert OTLP histograms.
+The `sum`, `count`, `min` and `max` within the OpenTelemetry histogram data are discarded for now until we support them in IntakeV2 histograms.
+
+## Labels
+
+Conceptually, elastic metric labels keys correspond to [OpenTelemetry Attributes](https://opentelemetry.io/docs/reference/specification/common/#attribute).
+When exporting, the agents MUST convert the attributes to labels as described in this section and MUST group metrics with the same attributes and OpenTelemetry instrumentation scope together into a single metricset.
+
+Attribute keys MUST be sanitized before being used as label keys: Every occurrence of the characters `.`, `*` and `"` must be replaced with a single `_`.
```

Comment:
[question] Is that something that is part of OTel specification or an additional constraint of our implementation ?

Reply:
Good question, I copied these sanitizations from our Micrometer implementation. In addition I validated that the OTLP intake does the same sanitizations. Maybe @felixbarny remembers where these restrictions come from? I don't think the otel-spec defines any character restrictions for attribute names, at least I wasn't able to find any.

Reply:
Hm, I don't quite remember where this came from. I could trace it back to https://github.com/elastic/apm-agent-java/commit/7265fe94 but unfortunately, I didn't document where this restriction comes from. Maybe these characters cause issues when trying to ingest or query for them in ES?

Reply:
OTel attribute name rules are here: https://opentelemetry.io/docs/reference/specification/common/attribute-naming/
They allow "any valid Unicode sequence", but limit to [U+0021 .. U+007E](https://en.wikipedia.org/wiki/Basic_Latin_%28Unicode_block%29#Table_of_characters) for semantic conventions. They explicitly suggest using '.' for namespacing, so our excluding '.' is a bit unfortunate.

APM server intake allows any unicode char **except `.`, `*`, and `"`**  -- https://github.com/elastic/apm-data/blob/3ad1a5caedc9bf5b36ebd62cdc8193e7cedc7d05/model/labels.go#L138-L164 -- with a comment that calls them "reserved characters".  I'm not sure why they are reserved. (Guessing: `.` reserved to avoid nesting under `labels.` causing mapping conflicts and template issues? `*` and `"` reserved to avoid quoting pain for querying labels?)

Given that APM server intake handles sanitizing label names, is there a desire/need/benefit to the APM agents doing so?

Reply:
APM server started doing the label sanitization in 7.11.0 (in https://github.com/elastic/apm-server/pull/4465), and in all 8.x releases.

Most APM agents technically supports back to APM server v6: https://www.elastic.co/guide/en/apm/guide/current/agent-server-compatibility.html

If the APM agents stop sanitizing label names, then sending a metricset with a label name with a `.`, for example, will be rejected/dropped by APM server 7.10 or earlier.  Does that mean that we should have APM agents continue to do this sanitization?

Options:
1. Continue to sanitize in the APM agents and add a note in the spec that this is only needed for APM server before 7.11.
2. Suggest that APM agents MAY skip sanitization if they know the APM server is >= v7.11.
3. Say that APM agents MAY skip sanitization, but warn users that OTel Metrics usage only works with APM server >=7.11, otherwise metricsets may get dropped.

I don't have a sense of actual usage of older APM server versions, or know its current support for these older versions.

Reply:
4. Only enable OTel metrics sending when connected to APM server >= 7.11 and never sanitize.

Overall, I think it's fine if OTel metrics sending doesn't work with EOL APM Servers.

Reply:
> 4. Only enable OTel metrics sending when connected to APM server >= 7.11 and never sanitize.

There might be a race in there. The agent will tend to have instrumented the OTel API or SDK used by the user's code before a round-trip to APM server to get its version. (At least for the Node.js agent. The agent *could* just no-op the exporting code, if the APM server turns out to be too old. An initial set of metricsets might get through.)

&lt;nobody-expects-the-spanish-inquisition>

5. Do NOT sanitize label names in the APM agents. Document that OTel Metrics shipping is only supported for APM server >=7.11, "for earlier APM server versions, metrics with label names including `.`, `*`, or `"` will get dropped".

</>

Reply:
> 5. Do NOT sanitize label names in the APM agents. Document that OTel Metrics shipping is only supported for APM server >=7.11, "for earlier APM server versions, metrics with label names including ., *, or " will get dropped".

That would also not be 100% correct, because IINM entire metricsets would be dropped, even if only one metric within them contains a dot.

I would just word documentation very generic: "OpenTelemetry Metrics support requires APM server 7.11 or newer".

Reply:
> because IINM entire metricsets would be dropped, even if only one metric within them contains a dot.

Metricsets are grouped by label sets, so it should just be as I described. However, the simpler generic support language you proposed is good.